src/Coordination/KeeperStorage.cpp

```diff
@@ -198,356 +176,650 @@ void KeeperStorage::Node::removeChild(StringRef child_path)
+        node->stat = committed_node.stat;
+        node->seq_num = committed_node.seq_num;
+        node->setData(committed_node.getData());
+    }
+
+    applyDeltas(
+        path,
+        Overloaded{
+            [&](const CreateNodeDelta & create_delta)
+            {
+                assert(!node);
+                node = std::make_shared<Node>();
+                node->stat = create_delta.stat;
+                node->setData(create_delta.data);
+            },
+            [&](const RemoveNodeDelta & /*remove_delta*/)
+            {
+                assert(node);
+                node = nullptr;
+            },
+            [&](const UpdateNodeDelta & update_delta)
+            {
+                assert(node);
+                update_delta.update_fn(*node);
+            },
+            [&](auto && /*delta*/) {},
+        });
+
+    return node;
+}
+
+bool KeeperStorage::UncommittedState::hasNode(StringRef path) const
+{
+    bool exists = storage.container.contains(std::string{path});
+    applyDeltas(
+        path,
+        Overloaded{
+            [&](const CreateNodeDelta & /*create_delta*/)
+            {
+                assert(!exists);
+                exists = true;
+            },
+            [&](const RemoveNodeDelta & /*remove_delta*/)
+            {
+                assert(exists);
+                exists = false;
+            },
+            [&](auto && /*delta*/) {},
+        });
+
+    return exists;
+}
+
+Coordination::ACLs KeeperStorage::UncommittedState::getACLs(StringRef path) const
+{
+    std::optional<uint64_t> acl_id;
+    if (auto maybe_node_it = storage.container.find(path); maybe_node_it != storage.container.end())
+        acl_id.emplace(maybe_node_it->value.acl_id);
+
+    const Coordination::ACLs * acls{nullptr};
+    applyDeltas(
+        path,
+        Overloaded{
+            [&](const CreateNodeDelta & create_delta)
+            {
+                assert(!acl_id);
+                acls = &create_delta.acls;
+            },
+            [&](const RemoveNodeDelta & /*remove_delta*/)
+            {
+                assert(acl_id || acls);
+                acl_id.reset();
+                acls = nullptr;
+            },
+            [&](const SetACLDelta & set_acl_delta)
+            {
+                assert(acl_id || acls);
+                acls = &set_acl_delta.acls;
+            },
+            [&](auto && /*delta*/) {},
+        });
+
+    if (acls)
+        return *acls;
+
+    return acl_id ? storage.acl_map.convertNumber(*acl_id) : Coordination::ACLs{};
+}
+
+namespace
+{
+
+[[noreturn]] void onStorageInconsistency()
+{
+    LOG_ERROR(&Poco::Logger::get("KeeperStorage"), "Inconsistency found between uncommitted and committed data. Keeper will terminate to avoid undefined behaviour.");
+    std::terminate();
+}
+
+}
+
+Coordination::Error KeeperStorage::commit(int64_t commit_zxid, int64_t session_id)
+{
+    for (auto & delta : uncommitted_state.deltas)
```

Comment:
Slightly unclear to me what will happen if don't have a delta corresponding to `commit_zxid`? Should we add an additional check for it?

Reply:
Deltas are added with increasing ZXIDs so if there doesn't exist any delta with the corresponding `commit_zxid` it will do nothing. 
For example, read requests don't generate any deltas.

Reply:
Maybe add exactly this message as comment to the code :)

Reply:
will do